### ISSUE TITLE
GC overhead limit exceeded

### ISSUE THREAD

**Issue by garypolhill** [NetLogo/NetLogo#1966]
TL;DR: 1. Something in my NetLogo code is causing Java garbage collection errors (and that's tricky to debug); 2. In headless mode, NetLogo ploughs on regardless when a fatal exception occurs, and probably it should just croak.

I get the following exception when running a model on an HPC cluster (i.e. using netlogo-headless.sh, but with `-XX:ParallelGCThreads=3` on the Java commandline to stop the run from spawning too many threads and hogging all the cores on the host:

```
Exception in thread "JobThread" java.lang.OutOfMemoryError: GC overhead limit exceeded
	at org.nlogo.agent.TurtleManagement.breedsOwnIndexOf(TurtleManagement.scala:47)
	at org.nlogo.agent.TurtleManagement.breedsOwnIndexOf$(TurtleManagement.scala:46)
	at org.nlogo.agent.World.breedsOwnIndexOf(World.scala:107)
	at org.nlogo.agent.Turtle.getBreedVariable(Turtle.java:308)
	at org.nlogo.prim._asm_procedureunhappyracism_report_0.perform(:4)
	at org.nlogo.nvm.Context.callReporterProcedure(Context.java:254)
	at org.nlogo.prim._asm_procedurehouseholdsaccumulateencounters_callreport_21.report(:1)
	at org.nlogo.nvm.Context.evaluateReporter(Context.java:239)
	at org.nlogo.prim._asm_procedurehouseholdsaccumulateencounters_ask_20.perform(:2)
	at org.nlogo.nvm.Context.runExclusive(Context.java:133)
	at org.nlogo.nvm.ExclusiveJob.run(ExclusiveJob.scala:39)
	at org.nlogo.nvm.Context.runExclusiveJob(Context.java:177)
	at org.nlogo.prim._asm_procedurehouseholdsaccumulateencounters_ask_14.perform(:1)
	at org.nlogo.nvm.Context.runExclusive(Context.java:133)
	at org.nlogo.nvm.ExclusiveJob.run(ExclusiveJob.scala:39)
	at org.nlogo.nvm.Context.runExclusiveJob(Context.java:177)
	at org.nlogo.prim._asm_procedurego_ask_37.perform(:1)
	at org.nlogo.nvm.Context.stepConcurrent(Context.java:107)
	at org.nlogo.nvm.ConcurrentJob.step(ConcurrentJob.scala:65)
	at org.nlogo.job.JobThread.runPrimaryJobs(JobThread.scala:133)
	at org.nlogo.job.JobThread.$anonfun$run$1(JobThread.scala:68)
	at org.nlogo.job.JobThread$$Lambda$53/0x0000000800d9fb00.apply$mcV$sp(Unknown Source)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at scala.util.control.Exception$Catch.apply(Exception.scala:228)
	at org.nlogo.api.Exceptions$.handling(Exceptions.scala:41)
	at org.nlogo.job.JobThread.run(JobThread.scala:66)
```

This is immediately followed in the output by this:

```
Run #1, RUNTIME ERROR: Reporter for measuring runs failed to report a result:
null
org.nlogo.lab.Worker$Runner$FailedException: Reporter for measuring runs failed to report a result:
null
	at org.nlogo.lab.Worker$Runner.$anonfun$callHelper$5(Worker.scala:155)
	at scala.collection.immutable.List.map(List.scala:293)
	at org.nlogo.lab.Worker$Runner.takeMeasurements$1(Worker.scala:151)
	at org.nlogo.lab.Worker$Runner.callHelper(Worker.scala:184)
	at org.nlogo.lab.Worker$Runner.call(Worker.scala:91)
	at org.nlogo.lab.Worker$Runner.call(Worker.scala:74)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

I suspect the latter error occurs because the former terminates `go` which means some housekeeping isn't done to provide one of the metrics from the experiments with data it needs (though surely it shouldn't return `null` even then...)

From [reading around](https://www.educative.io/edpresso/how-to-resolve-the-gc-overhead-limit-exceeded-error-in-java), it seems the garbage collector is hogging too many resources, suggesting I've written NetLogo code that leads to memory leaks. I'm guessing I shouldn't have to think about that when coding in NetLogo... For info, one of the two following procedures causes the offence; my money is on the latter, which creates a lot of temporary lists (which I've flagged in comments; see also some code in a [Stack Overflow answer](https://stackoverflow.com/questions/1393486/error-java-lang-outofmemoryerror-gc-overhead-limit-exceeded) on the topic):

```
to-report hdr
  if length htr > 0 [
    move-to first htr
    set htr but-first htr
  ]
  report length htr > 0
end

to hae
  let m-e h-e
  let e-h [h-e] of hhs-here ; *** list
  let h-n-e (length e-h) - 1
  let h-n-s-e (length filter [ eth -> eth = m-e ] e-h) - 1 ; *** list (and does filter only create one list?)
  let max-e modes e-h

  set hne hne + h-n-e
  set hsne hsne + h-n-s-e

  if (h-n-e > 0 
    and h-n-s-e < h-n-e
    and h-n-s-e / h-n-e < hh-h 
    and (member? m-e max-e or m-e = 1)
    and unh-h?) [

    ask one-of hhs-here with [h-e != m-e] [ 
      set unh-r lput ticks unh-r ; *** list
      if nw-unh? [
        ask stn-neighbors with [unh-r?] [
          set unh-r lput ticks unh-r ; *** list
        ]
      ]
      set hh-t hh-t - (r-d-t * hh-t)
    ]
  ]

  let f-here hhs-here with [member? myself stn-neighbors]
  if any? f-here [
    let ave-t mean [ hh-t ] of f-here
    let d-t ave-t - hh-t
    if abs d-t <= hh-tol [
      set hh-t hh-t + (vdt * d-t)
    ]
  ]
end
```

The following defines the variables and other ontology and procedures needed for the code to work:

```
globals [
  r-d-t ; number
  vdt ; number
  nw-unh? ; boolean
]

breed [hhs hh]
hhs-own [
  htr ; list
  h-e ; number
  hne ; number
  hnse ; number
  unh-r ; list
  unh-h ; list
  hh-h ; number
  hh-t ; number
  hh-tol ; number
  hh-tol-h ; number
  hh-tol-r ; number
]

undirected-link-breed [stns stn]

to-report unh-h?
  report length unh-h >= hh-tol-h
end

to-report unh-r?
  report length unh-r >= hh-tol-r
end
```
